File: Python/_438_Find_All_Anagrams_in_a_String/solution.py
class Solution:
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        res = []
        win = [0] * 26
        fixwin = [0] * 26
        
        if len(s) == 0 or len(p) == 0 or len(p) > len(s):
            return res
        
        for i in range(len(p)):
            fixwin[ord(p[i]) - ord('a')] += 1
        
        for i in range(len(p)):
            win[ord(s[i]) - ord('a')] += 1
        
        for i in range(len(s)):
            if self.isMatch(fixwin, win):
                res.append(i)
            if i + len(p) < len(s):
                win[ord(s[i]) - ord('a')] -= 1
                win[ord(s[i + len(p)]) - ord('a')] += 1
            else:
                break
        return res
    
    def isMatch(self, fixwin, win):
        for i in range(len(win)):
            if win[i] != fixwin[i]:
                return False
        return True

File: Java/_438_Find_All_Anagrams_in_a_String/Solution.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> res = new ArrayList<>();
        int[] win = new int[26];
        int[] fixwin = new int[26];

        if (s.length() == 0 || p.length() == 0 || p.length() > s.length())
            return res;

        for (int i = 0; i < p.length(); i++)
            fixwin[p.charAt(i) - 'a']++;

        for (int i = 0; i < p.length(); i++)
            win[s.charAt(i) - 'a']++;

        for (int i = 0; i < s.length(); i++) {
            if (isMatch(fixwin, win))
                res.add(i);
            
            if ((i + p.length()) < s.length()) {
                win[s.charAt(i) - 'a']--;
                win[s.charAt(i + p.length()) - 'a']++;
            } 
            else 
                break;
        }
        return res;
    }
    
    public boolean isMatch(int[] fixwin, int[] win) {
        for (int i = 0; i < win.length; i++)
            if (win[i] != fixwin[i])
                return false;
        return true;
    }
}